crates/nri-resctrl-plugin/src/lib.rs

```diff
@@ -0,0 +1,343 @@
+}
+
+/// Resctrl NRI plugin. Generic over `FsProvider` for testability.
+pub struct ResctrlPlugin<P: FsProvider = RealFs> {
+    cfg: ResctrlPluginConfig,
+    resctrl: Resctrl<P>,
+    state: Mutex<InnerState>,
+    tx: mpsc::Sender<PodResctrlEvent>,
+    dropped_events: Arc<AtomicUsize>,
+}
+
+impl ResctrlPlugin<RealFs> {
+    /// Create a new plugin with default real filesystem provider.
+    /// Returns the plugin and the receiver for emitted events.
+    pub fn new(cfg: ResctrlPluginConfig) -> (Self, mpsc::Receiver<PodResctrlEvent>) {
+        let (tx, rx) = mpsc::channel(cfg.event_channel_capacity);
+        let rc_cfg = ResctrlConfig {
+            group_prefix: cfg.group_prefix.clone(),
+            auto_mount: cfg.auto_mount,
+            ..Default::default()
+        };
+        let plugin = Self {
+            cfg,
+            resctrl: Resctrl::new(rc_cfg),
+            state: Mutex::new(InnerState::default()),
+            tx,
+            dropped_events: Arc::new(AtomicUsize::new(0)),
+        };
+        (plugin, rx)
+    }
+}
+
+impl<P: FsProvider> ResctrlPlugin<P> {
+    /// Create a new plugin with a custom resctrl handle (DI for tests).
+    /// Returns the plugin and the receiver for emitted events.
+    pub fn with_resctrl(
+        cfg: ResctrlPluginConfig,
+        resctrl: Resctrl<P>,
+    ) -> (Self, mpsc::Receiver<PodResctrlEvent>) {
+        let (tx, rx) = mpsc::channel(cfg.event_channel_capacity);
+        let plugin = Self {
+            cfg,
+            resctrl,
+            state: Mutex::new(InnerState::default()),
+            tx,
+            dropped_events: Arc::new(AtomicUsize::new(0)),
+        };
+        (plugin, rx)
+    }
+
+    /// Number of events dropped due to a full channel.
+    pub fn dropped_events(&self) -> usize {
+        self.dropped_events.load(Ordering::Relaxed)
+    }
+
+    /// Emit an event to the collector, drop if channel is full.
+    fn emit_event(&self, ev: PodResctrlEvent) {
+        if let Err(e) = self.tx.try_send(ev) {
+            self.dropped_events.fetch_add(1, Ordering::Relaxed);
+            warn!("resctrl-plugin: failed to send event: {}", e);
+        }
+    }
+}
+
+#[async_trait]
+impl<P: FsProvider + Send + Sync + 'static> Plugin for ResctrlPlugin<P> {
+    async fn configure(
+        &self,
+        _ctx: &TtrpcContext,
+        req: ConfigureRequest,
+    ) -> ttrpc::Result<ConfigureResponse> {
+        info!(
+            "Configured resctrl plugin for runtime: {} {}",
+            req.runtime_name, req.runtime_version
+        );
+
+        // Subscribe to container and pod lifecycle events.
+        let mut events = EventMask::new();
+        events.set(&[
+            Event::CREATE_CONTAINER,
+            Event::UPDATE_CONTAINER,
+            Event::STOP_CONTAINER,
+            Event::UPDATE_POD_SANDBOX,
+            Event::RUN_POD_SANDBOX,
+            Event::STOP_POD_SANDBOX,
+            Event::REMOVE_POD_SANDBOX,
+        ]);
```

Comment:
do we need all these events? I think CREATE_CONTAINER, RUN_POD_SANDBOX, and REMOVE_POD_SANDBOX should be good for what we need